Tom/foundation sdk docs (#106283)
grafana/grafana: docs/sources/observability-as-code/foundation-sdk/_index.md

- **Enhance version control:** Track changes seamlessly using standard version control systems like Git.
- **Automate deployments:** Integrate dashboard provisioning into your CI/CD pipelines for consistent and repeatable setups.

> The SDK supports multiple programming languages, including Go, TypeScript, Python, PHP, and Java, allowing you to choose the one that best fits your development environment.

Reviewer: ```suggestion
The SDK supports multiple programming languages, including Go, TypeScript, Python, PHP, and Java, so you can choose the one that best fits your development environment.
```

{{< youtube id="_OKQoABmg0Q" >}}

## Before you begin

Ensure you have the following prerequisites:

- **Programming environment:** Set up for your chosen language (e.g. Go, Node.js for TypeScript, Python 3.x for Python, etc).
- **Grafana instance:** A running Grafana instance compatible with the SDK version you’re using (refer to the [compatibility matrix](https://github.com/grafana/grafana-foundation-sdk#navigating-the-sdk)).

> - **Package manager:** Appropriate for your language (e.g. `npm` or `yarn` for TypeScript, `pip` for Python).

Reviewer: Same issue here.
```suggestion
- **Package manager:** Appropriate for your language, for example, `npm` or `yarn` for TypeScript or `pip` for Python.
```

> To get started quickly, clone the [intro-to-foundation-sdk repository](https://github.com/grafana/intro-to-foundation-sdk) which will give you complete examples along with a `docker-compose` stack.

Reviewer: ```suggestion
To get started,  clone the [intro-to-foundation-sdk repository](https://github.com/grafana/intro-to-foundation-sdk) to access examples and a `docker-compose` stack.
```

## Install the Grafana Foundation SDK

{{< code >}}
```go
go get github.com/grafana/grafana-foundation-sdk/go
```

```typescript
npm install @grafana/grafana-foundation-sdk
```
{{< /code >}}

For other languages, refer to the Grafana Foundation SDK documentation for detailed installation instructions.

## Grafana Foundation SDK Overview

> Before we jump into writing code, here's a quick overview of how the Grafana Foundation SDK works behind the scenes:

Reviewer: Rewording slightly to avoid colloquial expressions. 
```suggestion
Here's a quick overview of how the Grafana Foundation SDK works:
```

// Define a data source reference for our testdata data source
const testDataRef : dashboard.DataSourceRef = {
    type: 'grafana-testdata-datasource',
    uid:  'testdata',
};

// Define our dashboard as strongly typed code
const builder = new dashboard.DashboardBuilder('My Dashboard')
  .withPanel(
    new stat.PanelBuilder()
      .title('Version')
      .reduceOptions(
          new common.ReduceDataOptionsBuilder()
              .calcs(['lastNotNull'])
              .fields('/.*/')
      )
      .datasource(testdataRef)
      .withTarget(
          new testdata.DataqueryBuilder()
              .scenarioId('csv_content')
              .csvContent('version\nv1.2.3')
      )
  )
  .withPanel(
    new timeseries.PanelBuilder()
      .title('Random Time Series')
      .datasource(testdataRef)
      .withTarget(
        new testdata.DataqueryBuilder()
          .scenarioId('random_walk')
      )
  );

// Build the dashboard - errors in configuration will be thrown here
const dashboard = builder.build();

// Output the generated dashboard as JSON
console.log(JSON.stringify(dashboard, null, 2));
```

{{< /code >}}

> This code defines a dashboard titled “My Dashboard” with a two panels; a simple stat panel displaying a version number, and a time series panel displaying randomized data from the `testdata` data source `random_walk` scenario.

Reviewer: ```suggestion
This code defines a dashboard titled “My Dashboard” with a two panels:
* a simple stat panel displaying a version number, and 
* a time series panel displaying randomized data from the `testdata` data source `random_walk` scenario.
```

## Export and use the JSON

After you've defined your dashboard as code, build the final dashboard representation using the dashboard builder (typically using the `build()` function depending on language choice) and output the result as a JSON.

With the JSON payload you can:

- **Manually import:** Paste into Grafana’s dashboard import feature.
- **Automate:** Use [Grafana’s API](../../developers/http_api/) or the [Grafana CLI](../grafana-cli/) to programmatically upload the dashboard JSON.

> ## Concepts: Builders, Types, and Options

Reviewer: Do you need to list the concepts that you're discussing here or just say "concepts"?
```suggestion
## Concepts
```

Now that you've seen how to define a basic dashboard using code, let's take a moment to explain how it all works behind the scenes. The Grafana Foundation SDK is built around a few core concepts that make your dashboards structured, reusable, and strongly typed.

### Builders

> The SDK follows a builder pattern, which lets you compose dashboards step-by-step using chained method calls. Almost every piece of the dashboard (dashboards, panels, rows, queries, variables, etc) has its own `Builder` class.

Reviewer: ```suggestion
The SDK follows a builder pattern, which lets you compose dashboards step-by-step using chained method calls. 
Almost every piece of the dashboard, such as dashboards, panels, rows, queries,  and variables, has its own `Builder` class.
```
